README.md
script.myepisodecalendar
=================

XBMC plugin for [MyEpisodeCalendar](http://myepisodecalendar.com)

This plugin will set TV show episodes as watched on MyEpisodeCalendar.com.
New shows will also be added to your account if you start watching an episode.

Based on script.myepisodes by Maxime Hadjinlian.

build
=====
If you really want to build it, here is a simple script to do so:
```sh
#!/bin/bash

dest=script.myepisodecalendar
version=$(grep "^\s\+version" addon.xml | cut -f2 -d'"')

if [ -d $dest ]; then
    rm -r $dest
fi

mkdir $dest
cp addon.xml $dest/
cp *.txt $dest/
cp icon.png $dest/
cp *.py $dest/
cp -r resources $dest/

if [ -f $dest-$version.zip ]; then
    rm $dest-$version.zip
fi

zip -r $dest-$version.zip $dest
rm -r $dest
````
It will create a zip file that you can install directly within XBMC.

install
=======

Using the GUI of XBMC, choose to install your plugin as a zip file, find your
zip file, and you're done !

download
========
Get the latest release from the releases tab or download it [here](https://github.com/Connum/script.myepisodecalendar/releases/download/v1.0.2/script.myepisodecalendar-1.0.2.zip) directly.


changelog.txt
1.4.2
+ set python dependency version to recommended 2.24.0
+ removed empty tags in addon.xml > xbmc.addon.metadata to pass Travis schema validation
+ removed transparency in icon.png and minimized file size
+ removed script.tv.show.next.aired import left over from former integration
1.4.1
+ Fixed login by switching myepisodecalendar URLs to https
+ Switched TVDB API URLs to https
1.4.0
+ Hebrew translation thanks to Eng2Heb
+ major improvements in matching series titles
+ better performance searching for series titles
1.3.0
+ prevent some unhandled exceptions
+ cleaned up the code a bit
+ removed script.tv.show.next.aired integration (it is not desired by the Kodi team for addons to change other addons' settings)
1.2.1
+ Optimized the detection of metadata (TV show, season, episode)
+ Only run if detected media type is "tvshow", "season" or "episode" (or unset):
  This prevents the "XY not found" notification when watching movies.
1.2.0
+ Fixed problems with some show titles
+ Verified that the Add-On works as expected with Jarvis
+ Corrected wrong language string when show has been added automatically
+ Improved show search and show title mapping
1.1.1
+ Technical change only: use monitor.abortRequested() and monitor.waitForAbort() instead of xbmx.abortRequested and xbmx.sleep, see http://sourceforge.net/p/xbmc/mailman/message/34581714/
1.1.0
+ Added script.tv.show.next.aired integration:
  Series from MyEpisodeCalendar can now be added to the "TV Show - Next Aired" Add-On's setting "Extra shows". Series removed from MyEpisodeCalendar can also be automatically removed from Next Aired (changing this option requires saving the Add-On settings before taking effect when manually updating)
  Plus, MyEpisodeCalendar can automatically show the "Next Aired" GUI on startup.
  Known issue: MyEpisodeCalendar will now be listed in Programs, despite not doing more than just opening its settings dialogue.
1.0.2
+ Re-check login status and automatic re-login if required when marking an episode as watched.
  (myepisodecalendar.com seems to have a rather short time span for auto-logout)
1.0.1
+ Add German language
+ Some cosmetic changes for settings and notifications
+ New settings tab to disable/enable several notifications (errors will always be shown)

1.0.0 (not released) based on 1.0.15 of script.myepisodes by Maxime Hadjinlian.
======
1.0.15
+ Exclude pvr; http; https streaming.
+ Update BeautifulSoup to last version
+ Update Python to last version
+ Add missing license tag
1.0.14
+ Fix dummy mistake, sorry for the inconvenience
1.0.13
+ Update Regex expression to support more schemes.
+ Fixes bug when authentication failed
1.0.12
+ Improve show name detection again. Show like Archer or Revolution
should now work.
1.0.11
+ Improve the match between a title and MyEpisode's ID.
This should solve the issue where a show is not found and misfound when
its title contains a year, or country reference.
1.0.10
+ Fix a bug where Auto Add settings was not honored.
1.0.9
+ Fix a swapped assignment, episode was season and vice-versa.
1.0.8
+ Add an Auto Add shows settings.
+ Fix a memory leak
+ Fix some potential bug
1.0.7
+ Improve the way to find a TVShow title, thanks to trogdorsmith and
greek on the MyEpisodes forum for reporting the issue and testing "beta"
version.
+ Add a fallback to find a TVShow ID on MyEpisodes
1.0.6
+ Fix International issue. The plugin will now work when XBMC is set to
other lanuages than English.
+ Fix memory issue when exiting plugin.
1.0.5
+ Fix various bugs
+ Reload the plugin when the settings are modified
1.0.4
+ Fix impossibility to login
1.0.3
+ Update to new language system
1.0.2
+ Add Spanish (Thanks to stravieso)
1.0.1
+ Add Italian and Portuguese (Brazil) language, thanks
to Julia Vidile <julia.vidile@gmail.com>
1.0.0
* initial Frodo commit


default.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import re
import sys
import threading
import xbmc
import xbmcaddon

__addon__         = xbmcaddon.Addon()
__cwd__           = __addon__.getAddonInfo('path')
__icon__          = __addon__.getAddonInfo("icon")
__scriptname__    = __addon__.getAddonInfo('name')
__version__       = __addon__.getAddonInfo('version')
__language__      = __addon__.getLocalizedString
__resource_path__ = os.path.join(__cwd__, 'resources', 'lib')
__resource__      = xbmc.translatePath(__resource_path__).decode('utf-8')

from resources.lib.myepisodecalendar import MyEpisodeCalendar

class Monitor(xbmc.Monitor):
    def __init__( self, *args, **kwargs ):
        xbmc.Monitor.__init__( self )
        self.action = kwargs['action']

    def onSettingsChanged( self ):
        log('#DEBUG# onSettingsChanged')
        self.action()

class Player(xbmc.Player):

    def __init__ (self):
        xbmc.Player.__init__(self)
        log('Player - init')
        self.mye = self._loginMyEpisodeCalendar()
        if not self.mye.is_logged:
            return
        self.showid = self.episode = self.title = self.season = None
        self._totalTime = 999999
        self._lastPos = 0
        self._min_percent = int(__addon__.getSetting('watched-percent'))
        self._tracker = None
        self._playbackLock = threading.Event()
        self._monitor = Monitor(action = self._reset)

    def _reset(self):
        self._tearDown()
        if self.mye:
            del self.mye
        self.__init__()

    def _trackPosition(self):
        while self._playbackLock.isSet() and not xbmc.abortRequested:
            try:
                self._lastPos = self.getTime()
            except:
                self._playbackLock.clear()
            log('Inside Player. Tracker time = %s' % self._lastPos)
            xbmc.sleep(250)
        log('Position tracker ending with lastPos = %s' % self._lastPos)

    def _setUp(self):
        self._playbackLock.set()
        self._tracker = threading.Thread(target=self._trackPosition)

    def _tearDown(self):
        if hasattr(self, '_playbackLock'):
            self._playbackLock.clear()
        self._monitor = None
        if not hasattr(self, '_tracker'):
            return
        if self._tracker is None:
            return
        if self._tracker.isAlive():
            self._tracker.join()
        self._tracker = None

    def _loginMyEpisodeCalendar(self):
        username = __addon__.getSetting('Username')
        password = __addon__.getSetting('Password')

        login_notif = __language__(32912)
        if username is "" or password is "":
            notif(login_notif, time=2500)
            return None

        showLoginNotif = True
        mye = MyEpisodeCalendar(username, password)
        if mye.is_logged:
            if __addon__.getSetting('showNotif-login') != "true":
                showLoginNotif = False

            login_notif = "%s %s" % (username, __language__(32911))

        if showLoginNotif is True:
            notif(login_notif, time=2500)

        if mye.is_logged and (not mye.get_show_list()):
            notif(__language__(32927), time=2500)
        return mye

    def _mecReCheckAuth(self):
        if (not self.mye.is_logged):
            log("not logged in anymore")
            login_notif = __language__(32912)
            notif(login_notif, time=2500)
            return False
        return True

    def _addShow(self):
        # Add the show if it's not already in our account
        if self.showid in self.mye.shows.values():
            if __addon__.getSetting('showNotif-found') == "true":
                notif(self.title, time=2000)
            return
        was_added = self.mye.add_show(self.showid)
        added = 32926

        showAddNotif = True
        if was_added:
            if __addon__.getSetting('showNotif-autoadd') != "true":
                showAddNotif = False
            added = 32925
        
        if showAddNotif is True:
            notif("%s %s" % (self.title, __language__(added)))

    def onPlayBackStarted(self):
        self._setUp()
        self._totalTime = self.getTotalTime()
        self._tracker.start()

        filename_full_path = self.getPlayingFile().decode('utf-8')
        # We don't want to take care of any URL because we can't really gain
        # information from it.
        if _is_excluded(filename_full_path):
            self._tearDown()
            return

        # Try to find the title with the help of XBMC (Theses came from
        # XBMC.Subtitles add-ons)
        self.season = str(xbmc.getInfoLabel("VideoPlayer.Season"))
        log('Player - Season: %s' % self.season)
        self.episode = str(xbmc.getInfoLabel("VideoPlayer.Episode"))
        log('Player - Episode: %s' % self.episode)
        self.title = xbmc.getInfoLabel("VideoPlayer.TVshowtitle")
        log('Player - TVShow: %s' % self.title)
        if self.title == "":
            filename = os.path.basename(filename_full_path)
            log('Player - Filename: %s' % filename)
            self.title, self.season, self.episode = self.mye.get_info(filename)
            log('Player - TVShow: %s' % self.title)

        log("Title: %s - Season: %s - Ep: %s" % (self.title, self.season, self.episode))
        if not self.season and not self.episode:
            # It's not a show. If it should be recognised as one. Send a bug.
            self._tearDown()
            return

        self.showid = self.mye.find_show_id(self.title)
        if self.showid is None:
            notif("%s %s" % (self.title, __language__(32923)), time=3000)
            self._tearDown()
            return
        log('Player - Found : %s - %d (S%s E%s)' % (self.title,
                self.showid, self.season, self.episode))

        if __addon__.getSetting('auto-add') == "true":
            self._addShow()

    def onPlayBackStopped(self):
        # User stopped the playback
        self.onPlayBackEnded()

    def onPlayBackEnded(self):
        self._tearDown()

        actual_percent = (self._lastPos/self._totalTime)*100
        log('lastPos / totalTime : %s / %s = %s %%' % (self._lastPos,
            self._totalTime, actual_percent))
        if (actual_percent < self._min_percent):
            return

        # Playback is finished, set the items to watched
        found = 32923
        showMarkedNotif = True
        if self.mye.set_episode_watched(self.showid, self.season, self.episode):
            if __addon__.getSetting('showNotif-marked') != "true":
                showMarkedNotif = False
            found = 32924
        else:
            if (not self._mecReCheckAuth()):
                return False

        if showMarkedNotif is True:
            notif("%s (S%sE%s) %s" % (self.title, self.season.zfill(2), self.episode.zfill(2),
                __language__(found)))

def notif(msg, time=5000):
    notif_msg = "%s, %s, %i, %s" % ('MyEpisodeCalendar', msg.decode('utf-8', "replace"), time, __icon__)
    xbmc.executebuiltin("XBMC.Notification(%s)" % notif_msg.encode('utf-8'))

def log(msg):
    xbmc.log("### [%s] - %s" % (__scriptname__, msg.encode('utf-8'), ),
            level=xbmc.LOGDEBUG)

def _is_excluded(filename):
    log("_is_excluded(): Check if '%s' is a URL." % filename)
    excluded_protocols = ["pvr://", "http://", "https://"]
    return any(protocol in filename for protocol in excluded_protocols)

if ( __name__ == "__main__" ):
    player = Player()
    if not player.mye.is_logged:
        sys.exit(0)

    log( "[%s] - Version: %s Started" % (__scriptname__, __version__))

    while not xbmc.abortRequested:
        xbmc.sleep(100)

    player._tearDown()
    sys.exit(0)



resources/lib/myepisodecalendar.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from BeautifulSoup import BeautifulSoup
import cookielib
import re
import urllib, urllib2
 
# only for debugging
# import xbmc
# def log(msg):
#    xbmc.log("### [MyEpisodeCalendar] ### DEBUG: \n%s" % (msg, ),
#        level=xbmc.LOGDEBUG)

# This is totally stolen from script.xbmc.subtitles plugin !
REGEX_EXPRESSIONS = [
    '[Ss]([0-9]+)[][._-]*[Ee]([0-9]+)([^\\\\/]*)$',
    '[\._ \-]([0-9]+)x([0-9]+)([^\\/]*)',                     # foo.1x09
    '[\._ \-]([0-9]+)([0-9][0-9])([\._ \-][^\\/]*)',          # foo.109
    '([0-9]+)([0-9][0-9])([\._ \-][^\\/]*)',
    '[\\\\/\\._ -]([0-9]+)([0-9][0-9])[^\\/]*',
    'Season ([0-9]+) - Episode ([0-9]+)[^\\/]*',              # Season 01 - Episode 02
    'Season ([0-9]+) Episode ([0-9]+)[^\\/]*',                # Season 01 Episode 02
    '[\\\\/\\._ -][0]*([0-9]+)x[0]*([0-9]+)[^\\/]*',
    '[[Ss]([0-9]+)\]_\[[Ee]([0-9]+)([^\\/]*)',                #foo_[s01]_[e01]
    '[\._ \-][Ss]([0-9]+)[\.\-]?[Ee]([0-9]+)([^\\/]*)',       #foo, s01e01, foo.s01.e01, foo.s01-e01
    's([0-9]+)ep([0-9]+)[^\\/]*',                             #foo - s01ep03, foo - s1ep03
    '[Ss]([0-9]+)[][ ._-]*[Ee]([0-9]+)([^\\\\/]*)$',
    '[\\\\/\\._ \\[\\(-]([0-9]+)x([0-9]+)([^\\\\/]*)$'
    ]

MYEPISODE_URL = "http://www.myepisodecalendar.com"

def sanitize(title, replace):
    pattern = re.compile('[^A-Za-z0-9' + re.escape(replace) + ']', re.UNICODE | re.IGNORECASE)
    rPattern = re.compile('[\']+', re.UNICODE | re.IGNORECASE)
    sPattern = re.compile('\s+', re.UNICODE | re.IGNORECASE)

    title = title.replace('&', ' and ')
    title = title.replace(': ', ' - ')
    title = sPattern.sub(' ', title)

    title = pattern.sub(replace, title)
    title = rPattern.sub('', title)
    title = sPattern.sub(' ', title)

    return title.strip()

class MyEpisodeCalendar(object):

    def __init__(self, userid, password):
        self.userid = userid
        self.password = password
        self.shows = {}
        self.is_logged = False

        self.init_request_opener()
        self.do_login()

    def init_request_opener(self):
        self.cj = cookielib.CookieJar()
        self.opener = urllib2.build_opener(
            urllib2.HTTPRedirectHandler(),
            urllib2.HTTPHandler(debuglevel=0),
            urllib2.HTTPSHandler(debuglevel=0),
            urllib2.HTTPCookieProcessor(self.cj)
        )
        self.opener.addheaders = [
            ('User-agent', 'Lynx/2.8.1pre.9 libwww-FM/2.14')
        ]

    # force logout, mainly for debugging
    def do_logout(self):
        self.init_request_opener()

    def do_login(self):
        login_data = urllib.urlencode({
            'email' : self.userid,
            'password' : self.password,
            'submit' : "Login",
            })
        login_url = "%s/%s" % (MYEPISODE_URL, "login/")
        self.send_req(login_url, login_data)

        return self.check_login_status_from_settings()

    def check_login_status_from_settings(self):
        settings_url = "%s/%s" % (MYEPISODE_URL, "settings/")
        data = self.send_req(settings_url)

        self.is_logged = True

        # Quickly check if it seems we are logged on.
        if (data is None) or (self.userid not in data):
            self.is_logged = False

        return self.is_logged

    def check_login_status_from_data(self, data):
        self.is_logged = True

        # Quickly check if it seems we are logged on.
        if (data is None) or (">Logout<" not in data):
            self.is_logged = False

        return self.is_logged

    def send_req(self, url, data = None):
        try:
            response = self.opener.open(url, data)
            return ''.join(response.readlines())
        except:
            return None

    def get_show_list(self):
        # Populate shows with the list of show_ids in our account
        myshows_url = "%s/%s" % (MYEPISODE_URL, "myShows")
        data = self.send_req(myshows_url)


        if data is None:
            return False
        soup = BeautifulSoup(data)
        mylist = soup.find("div", {"class": re.compile(r'\bpageRowContainer\b')})
        mylist_tr = mylist.findAll("div", {"class": re.compile(r'\bbigBannerContainer\b')})

        for row in mylist_tr:
            link = row.find('a', {'href': True})
            link_url = link.get('href')
            showid = link_url.split('/')[-1]
            key = link.get('title').strip()
            sanitized_key = sanitize(key, ' ')
            if sanitized_key != key:
                key = ";".join([key, sanitized_key])
            self.shows[key] = int(showid)
        return True

    def find_show_link(self, data, show_name, strict=False):

        if data is None:
            return None, None

        soup = BeautifulSoup(data)
        tvshownames = soup.findAll("div", {"class": re.compile(r'\btvShowName\b')})

        show_href = None
        show_name_sanitized = sanitize(show_name.lower(), ' ')
        # TODO: search for dashed notation in URL alternatively
        # show_name_dashes = sanitize(show_name.lower(), '-') # TODO: different sanitization: reduce multiple dashes to one, remove single quote...

        for show in tvshownames:
            link = show.find("a", href=True)
            if link.string is None:
                continue
            if strict:
                if sanitize(link.string.lower(), ' ') == show_name_sanitized:
                    show_href = link.get('href')
                    show_name = link.get('title')
                    break
            else:
                if sanitize(link.string.lower(), ' ').startswith(show_name_sanitized):
                    show_href = link.get('href')
                    show_name = link.get('title')
                    break
        return show_name, show_href

    def find_show_id(self, show_name):
        # Try to find the ID of the show in our account first
        # Create a slice with only the show that may match
        new_name = show_name
        show_name = show_name.lower().strip()
        show_name_sanitized = sanitize(show_name, ' ')

        for keys, v in self.shows.iteritems():
            if ';' in keys:
                keys = keys.split(';')
            else:
                keys = [keys,]

            for k in keys:
                k_lower = k.lower()
                if show_name == k_lower or show_name_sanitized == k_lower:
                    new_name = keys[0]
                    return new_name, v


        # It's not in our account yet ?
        # Try Find a show through its name and report its id
        search_url = "%s/%s/%s" % (MYEPISODE_URL, "search", sanitize(show_name.strip(' .'), '_'))
        data = self.send_req(search_url)
        new_name, show_href = self.find_show_link(data, show_name)

        # TODO: try to automatically request a show to be added to MEC
        # TODO: check if this is working for MEC
        if show_href is None:
            # Try to lookup the list of all the shows to find the exact title
            list_url = "%s/%s/%s" % (MYEPISODE_URL, "allShows/filter-by-name",
                    show_name[0].upper())
            data = self.send_req(list_url)
            new_name, show_href = self.find_show_link(data, show_name, strict=True)

        # Really did not find anything :'(
        if show_href is None:
            return new_name, None

        showid = show_href.split('/')[-1]

        if showid is None:
            return new_name, None

        showid = int(showid.strip())

        return new_name, showid

    # This is totally stolen from script.xbmc.subtitles plugin !
    def get_info(self, file_name):
        title = None
        episode = None
        season = None
        for regex in REGEX_EXPRESSIONS:
            response_file = re.findall(regex, file_name)
            if len(response_file) > 0 :
                season = response_file[0][0]
                episode = response_file[0][1]
            else:
                continue
            title = re.split(regex, file_name)[0]
            title = sanitize(title, ' ')
            title = title.strip()
            return title.title(), season, episode
        return None, None, None

    def add_show(self, show_id):
        # Try to add the show to your account.
        tracklist_data = urllib.urlencode({
            'tvShowID' : show_id,
            'type' : "add",
            })
        url = "%s/actions/trackList.php" % MYEPISODE_URL
        data = self.send_req(url, tracklist_data)
        if data is None:
            return False
        # Update list
        # TODO: instead of calling get_show_list(),
        #       which takes another request and updates all shows unneccessarily,
        #       just add this show's data!
        self.get_show_list()
        return True

    def set_episode_watched(self, show_id, season, episode):
        season_url = "%s/episodes//%d/Season-%d" % (MYEPISODE_URL,
                show_id, int(season))
        episode = int(episode)

        season_data = self.send_req(season_url)

        # re-login if logged out meanwhile
        if (not self.check_login_status_from_data(season_data)):
            if self.do_login():
                season_data = self.send_req(season_url)
            else:
                return False

        soup = BeautifulSoup(season_data)

        divs_epno = soup.findAll("div", {"id": "episodeNo"})
        if divs_epno is None:
            return False

        for epnodiv in divs_epno:
            try:
                intContent = int(epnodiv.string.strip())
            except:
                continue
            else:
                if intContent == int(episode):
                    div_epno = epnodiv
                    break


        # episode could not be found on this page...
        if div_epno is None:
            return False

        episode_id = re.findall('[^\d](\d+)$', div_epno.parent.get("id"))[0]

        episode_add_data = urllib.urlencode({
            'episodeID' : episode_id,
            'season' : int(season),
            'tvShowID' : show_id,
            'type' : "add",
            })
        seen_url = "%s/actions/watchList.php" % MYEPISODE_URL
        data = self.send_req(seen_url, episode_add_data)

        if data is None:
            return False

        return True

    def getTVDBIDFromShowTitle(self, title, lang='en'):
        GetSeriesURL = 'http://www.thetvdb.com/api/GetSeries.php?seriesname=%s&language=%s' % (urllib.quote(title), lang)
        seriesData = self.send_req(GetSeriesURL)

        if seriesData is None:
            return False

        soup = BeautifulSoup(seriesData)

        try:
            seriesid = soup.find('seriesid').text;
        except:
            seriesid = False

        return seriesid


main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import re
import sys
import threading
import xbmc, xbmcaddon

__addon__         = xbmcaddon.Addon()
__addonid__       = __addon__.getAddonInfo('id')
__cwd__           = __addon__.getAddonInfo('path')
__icon__          = __addon__.getAddonInfo("icon")
__scriptname__    = __addon__.getAddonInfo('name')
__version__       = __addon__.getAddonInfo('version')
__language__      = __addon__.getLocalizedString
__resource_path__ = os.path.join(__cwd__, 'resources', 'lib')
__resource__      = xbmc.translatePath(__resource_path__).decode('utf-8')
__datapath__      = os.path.join(xbmc.translatePath('special://masterprofile/addon_data/').decode('utf-8'), __addonid__)


from resources.lib.myepisodecalendar import MyEpisodeCalendar

class Monitor(xbmc.Monitor):
    def __init__( self, *args, **kwargs ):
        xbmc.Monitor.__init__( self )
        self.action = kwargs['action']

    def onSettingsChanged( self ):
        log('#DEBUG# onSettingsChanged')
        self.action()

class Player(xbmc.Player):

    def __init__ (self):
        xbmc.Player.__init__(self)
        log('Player - init')
        self.mye = self._loginMyEpisodeCalendar()
        if not self.mye.is_logged:
            return
        self.showid = self.episode = self.title = self.season = None
        self._totalTime = 999999
        self._lastPos = 0
        self._min_percent = int(__addon__.getSetting('watched-percent'))
        self._tracker = None
        self._playbackLock = threading.Event()
        self._monitor = Monitor(action = self._reset)

    def _reset(self):
        self._tearDown()
        if self.mye:
            del self.mye
        self.__init__()

    def _trackPosition(self):
        while self._playbackLock.isSet() and not xbmc.abortRequested:
            try:
                self._lastPos = self.getTime()
            except:
                self._playbackLock.clear()
            log('Inside Player. Tracker time = %s' % self._lastPos)
            xbmc.sleep(250)
        log('Position tracker ending with lastPos = %s' % self._lastPos)

    def _setUp(self):
        self._playbackLock.set()
        self._tracker = threading.Thread(target=self._trackPosition)

    def _tearDown(self):
        if hasattr(self, '_playbackLock'):
            self._playbackLock.clear()
        self._monitor = None
        if not hasattr(self, '_tracker'):
            return
        if self._tracker is None:
            return
        if self._tracker.isAlive():
            self._tracker.join()
        self._tracker = None

    @staticmethod
    def _loginMyEpisodeCalendar(silent=False):
        username = __addon__.getSetting('Username')
        password = __addon__.getSetting('Password')

        login_notif = __language__(32912)
        if username is "" or password is "":
            notif(login_notif, time=2500)
            return None

        showLoginNotif = True
        mye = MyEpisodeCalendar(username, password)
        if mye.is_logged:
            if __addon__.getSetting('showNotif-login') != "true" or silent:
                showLoginNotif = False

            login_notif = "%s %s" % (username, __language__(32911))

        if showLoginNotif is True:
            notif(login_notif, time=2500)

        if mye.is_logged and (not mye.get_show_list()):
            notif(__language__(32927), time=2500)

        return mye

    def _mecReCheckAuth(self):
        if (not self.mye.is_logged):
            log("not logged in anymore")
            login_notif = __language__(32912)
            notif(login_notif, time=2500)
            return False
        return True

    def _addShow(self):
        # Add the show if it's not already in our account
        if self.showid in self.mye.shows.values():
            if __addon__.getSetting('showNotif-found') == "true":
                notif(self.title, time=2000)
            return
        was_added = self.mye.add_show(self.showid)
        addedStringId = 32926

        showAddNotif = True
        if was_added:
            if __addon__.getSetting('showNotif-autoadd') != "true":
                showAddNotif = False
            addedStringId = 32925

        if showAddNotif is True:
            notif("%s %s" % (self.title, __language__(addedStringId)))

    def onPlayBackStarted(self):
        self._setUp()
        self._totalTime = self.getTotalTime()
        self._tracker.start()

        filename_full_path = self.getPlayingFile().decode('utf-8')

        playlist = xbmc.PlayList(xbmc.PLAYLIST_VIDEO)

        if playlist.size() < 1:
            log('Playlist empty - this is probably a PVR channel. Abort.')
            self._tearDown()
            return

        currentPlaylistItemInfo = playlist[playlist.getposition()].getVideoInfoTag()

        mediaType = xbmc.getInfoLabel("ListItem.DBType") or currentPlaylistItemInfo.getMediaType()
        log('Player - MediaType: %s' % mediaType)

        # We don't want to take care of any URL because we can't really gain
        # information from it. We also only want to check TV episodes, so skip everything else.
        # (not explicitely checking for "episode", "season" or "tvshow", because sometimes it may return empty as well)
        if mediaType in ["movie", "musicvideo", "music", "song", "album", "artist"] or _is_excluded(filename_full_path):
            self._tearDown()
            return

        # Try to find the title with the help of XBMC (Theses came from
        # XBMC.Subtitles add-ons)
        # Falling back to currentPlaylistItemInfo, because getInfoLabel() seems to return empty values most of the time!
        self.season = str(xbmc.getInfoLabel("VideoPlayer.Season")) or currentPlaylistItemInfo.getSeason()
        log('Player - Season: %s' % self.season)
        self.episode = str(xbmc.getInfoLabel("VideoPlayer.Episode")) or currentPlaylistItemInfo.getEpisode()
        log('Player - Episode: %s' % self.episode)
        self.title = xbmc.getInfoLabel("VideoPlayer.TVshowtitle") or currentPlaylistItemInfo.getTVShowTitle()
        log('Player - TVShow: %s' % self.title)
        if self.title == "":
            filename = os.path.basename(filename_full_path)
            log('Player - Filename: %s' % filename)
            self.title, self.season, self.episode = self.mye.get_info(filename)
            log('Player - TVShow: %s' % self.title)

        log("Title: %s - Season: %s - Ep: %s" % (self.title, self.season, self.episode))
        if not self.season and not self.episode:
            # It's not a show. If it should be recognised as one. Send a bug.
            self._tearDown()
            return

        self.title, self.showid = self.mye.find_show_id(self.title)
        if self.showid is None:
            notif("%s %s" % (self.title, __language__(32923)), time=3000)
            self._tearDown()
            return
        log('Player - Found : %s - %d (S%s E%s)' % (self.title,
                self.showid, self.season, self.episode))

        if __addon__.getSetting('auto-add') == "true":
            self._addShow()

    def onPlayBackStopped(self):
        # User stopped the playback
        self.onPlayBackEnded()

    def onPlayBackEnded(self):
        self._tearDown()

        if self._totalTime == 0 or self._totalTime is None:
            log('could not get totalTime - assume finished')
            actual_percent = 100
        else: 
            actual_percent = (self._lastPos/self._totalTime)*100
            log('lastPos / totalTime : %s / %s = %s %%' % (self._lastPos,
                self._totalTime, actual_percent))
        if actual_percent < self._min_percent or self.showid is None or self.season is None or self.episode is None:
            return

        # Playback is finished, set the items to watched
        found = 32923
        showMarkedNotif = True
        if self.mye.set_episode_watched(self.showid, self.season, self.episode):
            if __addon__.getSetting('showNotif-marked') != "true":
                showMarkedNotif = False
            found = 32924
        else:
            if (not self._mecReCheckAuth()):
                return False

        if showMarkedNotif is True:
            notif("%s (S%sE%s) %s" % (self.title, self.season.zfill(2), self.episode.zfill(2),
                __language__(found)))

def notif(msg, time=5000):
    notif_msg = "\"%s\", \"%s\", %i, %s" % ('MyEpisodeCalendar', msg.decode('utf-8', "replace"), time, __icon__)
    xbmc.executebuiltin("XBMC.Notification(%s)" % notif_msg.encode('utf-8'))

def log(msg):
    try:
        msg = msg.encode('utf-8')
    except:
        pass

    xbmc.log("### [%s] - %s" % (__scriptname__, msg, ),
            level=xbmc.LOGDEBUG)

def _is_excluded(filename):
    log("_is_excluded(): Check if '%s' is a URL." % filename)
    excluded_protocols = ["pvr://", "http://", "https://"]
    return any(protocol in filename for protocol in excluded_protocols)


if ( __name__ == "__main__" ):
    monitor = xbmc.Monitor()
    player = Player()
    if not player.mye.is_logged:
        sys.exit(0)

    log( "[%s] - Version: %s Started" % (__scriptname__, __version__))

    while not monitor.abortRequested():
        if monitor.waitForAbort(10):
            break

    player._tearDown()
    sys.exit(0)